Mathlib/Data/Set/Intervals/Instances.lean

```diff
@@ -0,0 +1,396 @@
+
+@[simp]
+theorem mk_one [Nontrivial α] (h : (1 : α) ∈ Ioc (0 : α) 1) : (⟨1, h⟩ : Ioc (0 : α) 1) = 1 :=
+  rfl
+#align set.Ioc.mk_one Set.Ioc.mk_one
+
+@[simp, norm_cast]
+theorem coe_eq_one [Nontrivial α] {x : Ioc (0 : α) 1} : (x : α) = 1 ↔ x = 1 := by
+  symm
+  exact Subtype.ext_iff
+#align set.Ioc.coe_eq_one Set.Ioc.coe_eq_one
+
+theorem coe_ne_one [Nontrivial α] {x : Ioc (0 : α) 1} : (x : α) ≠ 1 ↔ x ≠ 1 :=
+  not_iff_not.mpr coe_eq_one
+#align set.Ioc.coe_ne_one Set.Ioc.coe_ne_one
+
+theorem coe_pos (x : Ioc (0 : α) 1) : 0 < (x : α) :=
+  x.2.1
+#align set.Ioc.coe_pos Set.Ioc.coe_pos
+
+theorem coe_le_one (x : Ioc (0 : α) 1) : (x : α) ≤ 1 :=
+  x.2.2
+#align set.Ioc.coe_le_one Set.Ioc.coe_le_one
+
+/-- like `coe_le_one`, but with the inequality in `Ioc (0:α) 1`. -/
+theorem le_one [Nontrivial α] {t : Ioc (0 : α) 1} : t ≤ 1 :=
+  t.2.2
+#align set.Ioc.le_one Set.Ioc.le_one
+
+instance mul : Mul (Ioc (0 : α) 1) where
+  mul p q := ⟨p.1 * q.1, ⟨mul_pos p.2.1 q.2.1, mul_le_one p.2.2 (le_of_lt q.2.1) q.2.2⟩⟩
+#align set.Ioc.has_mul Set.Ioc.mul
+
+instance pow : Pow (Ioc (0 : α) 1) ℕ where
+  pow p n := ⟨p.1 ^ n, ⟨pow_pos p.2.1 n, pow_le_one n (le_of_lt p.2.1) p.2.2⟩⟩
+#align set.Ioc.has_pow Set.Ioc.pow
+
+@[simp, norm_cast]
+theorem coe_mul (x y : Ioc (0 : α) 1) : ↑(x * y) = (x * y : α) :=
+  rfl
+#align set.Ioc.coe_mul Set.Ioc.coe_mul
+
+@[simp, norm_cast]
+theorem coe_pow (x : Ioc (0 : α) 1) (n : ℕ) : ↑(x ^ n) = ((x : α) ^ n) :=
+  rfl
+#align set.Ioc.coe_pow Set.Ioc.coe_pow
+
+instance semigroup : Semigroup (Ioc (0 : α) 1) :=
+  Subtype.coe_injective.semigroup _ coe_mul
+#align set.Ioc.semigroup Set.Ioc.semigroup
+
+instance monoid [Nontrivial α] : Monoid (Ioc (0 : α) 1) :=
+  Subtype.coe_injective.monoid _ coe_one coe_mul coe_pow
+#align set.Ioc.monoid Set.Ioc.monoid
+
+instance commSemigroup {α : Type _} [StrictOrderedCommSemiring α] : CommSemigroup (Ioc (0 : α) 1) :=
+  Subtype.coe_injective.commSemigroup _ coe_mul
+#align set.Ioc.comm_semigroup Set.Ioc.commSemigroup
+
+instance commMonoid {α : Type _} [StrictOrderedCommSemiring α] [Nontrivial α] :
+    CommMonoid (Ioc (0 : α) 1) :=
+  Subtype.coe_injective.commMonoid _ coe_one coe_mul coe_pow
+#align set.Ioc.comm_monoid Set.Ioc.commMonoid
+
+instance cancelMonoid {α : Type _} [StrictOrderedRing α] [IsDomain α] :
+    CancelMonoid (Ioc (0 : α) 1) :=
+  { Set.Ioc.monoid with
+    mul_left_cancel := fun a _ _ h =>
+      Subtype.ext <| mul_left_cancel₀ a.prop.1.ne' <| (congr_arg Subtype.val h : _)
+    mul_right_cancel := fun _ b _ h =>
+      Subtype.ext <| mul_right_cancel₀ b.prop.1.ne' <| (congr_arg Subtype.val h : _) }
+#align set.Ioc.cancel_monoid Set.Ioc.cancelMonoid
+
+section
+-- Porting note: This takes too long
+set_option maxHeartbeats 0
+
+instance cancelCommMonoid {α : Type _} [StrictOrderedCommRing α] [IsDomain α] :
+    CancelCommMonoid (Ioc (0 : α) 1) :=
+  { Set.Ioc.cancelMonoid, Set.Ioc.commMonoid with }
+#align set.Ioc.cancel_comm_monoid Set.Ioc.cancelCommMonoid
```

Comment:
I guess this is the only blocker at this point, though we've landed a few of these already.